### PR TITLE
Add noetic-alpine3.14

### LINKED ISSUE #454: Support newer Alpine Linux
v3.11 will reach EOL at Nov 1, 2021

### PR DESCRIPTION
**Pull request by at-wat**
Close #454

v3.11 will reach EOL at Nov 1, 2021.
v3.14 EOL will be May 1, 2023.

## Notable changes on noetic-alpine3.14

- Entirely drop py2 dependencies
  - `backports/py-*` is replaced by `backports/py3-*`
- Enable stack protector
  - (It was disabled during the first attempt to port ROS to Alpine to fix large performance degradation, but stack-protector was not the cause)
- Suppress some compiler warnings
  - Include path of `fcntl.h` and `poll.h`
  - Boost bind placeholder namespace
- Switch to OpenSSL from libressl
  - Alpine has been switched back to OpenSSL since 3.9: https://lists.alpinelinux.org/~alpine/devel/%3C20181011171746.4c01f758%40ncopa-desktop.copa.dup.pw%3E
  - rosdep key is updated on 3.14 to resolve conflicts https://github.com/alpine-ros/rosdistro/pull/87

### CODE REVIEW

**Comment by at-wat** [comment]
waiting Alpine 3.14

**Comment by at-wat** [comment]
Seems be failing due to seccomp problem.

**Comment by at-wat** [comment]
```
during GIMPLE pass: strlen
In file included from /home/builder/aports/ros/noetic/ros-noetic-compressed-image-transport/abuild/src/compressed_image_transport/include/compressed_image_transport/compressed_subscriber.h:37,
                 from /home/builder/aports/ros/noetic/ros-noetic-compressed-image-transport/abuild/src/compressed_image_transport/src/compressed_subscriber.cpp:35:
/usr/ros/noetic/include/dynamic_reconfigure/server.h: In member function 'bool dynamic_reconfigure::Server<ConfigType>::setConfigCallback(dynamic_reconfigure::Reconfigure::Request&, dynamic_reconfigure::Reconfigure::Response&) [with ConfigType = compressed_image_transport::CompressedSubscriberConfig]':
/usr/ros/noetic/include/dynamic_reconfigure/server.h:213:8: internal compiler error: Segmentation fault
  213 |   bool setConfigCallback(dynamic_reconfigure::Reconfigure::Request &req,
      |        ^~~~~~~~~~~~~~~~~
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://gcc.gnu.org/bugs/> for instructions.
```

File: ros/noetic.v3.14/ros-noetic-roscpp/workaround-gcc-segv.patch [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
++++ b/include/ros/subscription_callback_helper.h
+@@ -123,7 +123,7 @@ public:
+ 
+     if (!msg)
+     {
+-      ROS_DEBUG("Allocation failed for message of type [%s]", getTypeInfo().name());
++      // ROS_DEBUG("Allocation failed for message of type [%s]", getTypeInfo().name());
+       return VoidConstPtr();
+     }
+
```
Comment (at-wat) [inline ros/noetic.v3.14/ros-noetic-roscpp/workaround-gcc-segv.patch:47]: workaround for gcc internal compiler error
https://github.com/seqsense/aports-ros-experimental/pull/504#issuecomment-863246065

**Comment by at-wat** [comment]
@f-fl0 please take a look

**Comment by at-wat** [comment]
noetic-3.11 makedepends to `py-` are also replaced by `py3-` by https://github.com/alpine-ros/ros-abuild-docker/pull/135

File: backports/py3-rosdistro/force-to-enable-alpine-platform.patch [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+                     self.release_platforms[os_name].append(os_code_name)
+ 
++        # Tweak for alpine platform
++        self.release_platforms['alpine'] = []
++        if name == 'kinetic':
++            self.release_platforms['alpine'].append('3.7')
++        if name == 'melodic':
++            self.release_platforms['alpine'].append('3.8')
++        if name == 'noetic':
++            self.release_platforms['alpine'].append('3.11')
```
Comment (f-fl0) [inline backports/py3-rosdistro/force-to-enable-alpine-platform.patch:16]: Should we also add `3.14` for noetic?

File: backports/py3-rosdistro/force-to-enable-alpine-platform.patch [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+--- a/src/rosdistro/distribution_file.py
++++ b/src/rosdistro/distribution_file.py
+@@ -74,6 +74,15 @@ class DistributionFile(object):
+                     assert os_code_name not in self.release_platforms[os_name], "Distribution '%s' specifies the os_code_name '%s' multiple times for the os_name '%s'" % (self.name, os_code_name, os_name)
+                     self.release_platforms[os_name].append(os_code_name)
+ 
++        # Tweak for alpine platform
++        self.release_platforms['alpine'] = []
++        if name == 'kinetic':
++            self.release_platforms['alpine'].append('3.7')
```
Comment (f-fl0) [inline backports/py3-rosdistro/force-to-enable-alpine-platform.patch:12]: `kinetic` can probably be removed.

File: Makefile [hunk 4]
```diff
@@ -2,13 +2,20 @@ SHELL                   = /bin/bash
 
 ROS_DISTRO             ?= noetic
 BUILDER_NAME            = seqsense/aports-ros-builder
-ALPINE_VERSION          = $(shell . ./alpine_version_from_ros_distro.sh; alpine_version_from_ros_distro $(ROS_DISTRO))
+ALPINE_VERSION          = 3.11
```
Comment (f-fl0) [inline Makefile:5]: Should it be `ALPINE_VERSION          ?= 3.11` so that `ALPINE_VERSION` can be set to `3.14`?

**Review by f-fl0 — COMMENTED** [review]
Thanks for this massive work. I didn't check all the packages in details as there are too many and I assume a lot of the modifications are due to `py => py3`.

I noticed a few mention to `v3.7`. They are probably remains of https://github.com/seqsense/aports-ros-experimental/pull/496. Is it necessary to still support it?

File: backports/py3-rosdistro/force-to-enable-alpine-platform.patch (hunk 3 above)
Comment (at-wat) [reply backports/py3-rosdistro/force-to-enable-alpine-platform.patch:12]: I think name resolution is better to be remained.

File: Makefile (hunk 4 above)
Comment (at-wat) [reply Makefile]: `ALPINE_VERSION = 3.11` can be overwritten by `make ALPINE_VERSION=3.14` as well.
But `?=` may be better for readability to mark it is configurable.

**Review by f-fl0 — APPROVED** [review]
LGTM. If necessary, removing any mention to `3.7` can be done in a separate PR.

**Pull request merged**